Refactor DistanceExt to use generic traits with comprehensive cross-type support and Bugfix

zhangfengcdt opened the pull request:
## Summary
  - Refactor `DistanceExt` implementation to use generic programming approach more
  - Add comprehensive cross-type distance calculations for all geometry combinations
  - Implement systematic macro-based generation for polygon-like and multi-geometry types
  - Extract utility functions to separate module for better organization and reusability
  - **Cross-validation tests** against concrete implementations (found and fixed bugs)

## Key Changes
  - **Generic Trait Implementation**: Transform distance calculations to work with any WKB-compatible geometry type through trait bounds
  - **Cross-Type Support**: Enable distance calculations between different geometry types (Point-to-Polygon, LineString-to-MultiPolygon,  etc.)
  - **Utility Module**: Extract core distance functions (`point_distance_generic`, `line_segment_distance_generic`,  `nearest_neighbour_distance`, etc.) to `utils.rs`
  - **Macro System**: Implement systematic macro-based code generation for polygon-like geometries and multi-geometry collections
  - **Symmetric Operations**: Add proper symmetric distance implementations using macro-generated code
  - **Performance**: Maintain performance parity with concrete implementations while adding generic flexibility

## Tests Coverage

The comprehensive test suite contains 76 tests organized into the following categories. All tests validate that generic trait-based implementations produce identical results to concrete implementations, ensuring mathematical correctness and cross-compatibility across different geometry type systems.

  1. Basic Function Tests (8 tests)

  - Point distance calculations (basic, same point, negative coordinates, empty points)
  - Line segment distance calculations (perpendicular, diagonal, on-line, zero-length, before/beyond endpoints)

  2. Core Distance Algorithm Tests (4 tests)

  - nearest_neighbour_distance function (basic, intersecting, parallel lines, single segments)

  3. Geometry Utility Function Tests (4 tests)

  - ring_contains_coord function (inside, outside, on boundary, triangle shapes)

  4. Cross-Type Distance Function Tests (15 tests)

  - Point-to-LineString distance (basic, empty, single point, on linestring)
  - Point-to-Polygon distance (inside, outside, on boundary, with holes, empty)
  - LineString-to-Polygon distance (outside, intersecting, empty polygon)
  - Line-to-Line distance (parallel, intersecting, skew)
  - Polygon-to-Polygon distance (separate, intersecting, containment scenarios)
  - Line-to-LineString and Line-to-Polygon distance
  - Triangle-to-Point distance (inside, outside)

  5. Symmetry Property Tests (3 tests)

  - Point-LineString symmetry validation
  - Point-Polygon symmetry validation
  - LineString-Polygon symmetry validation

  6. Edge Cases and Degenerate Geometry Tests (4 tests)

  - Empty geometries handling
  - Single-point LineStrings
  - Identical point LineStrings
  - Cross-validation with existing concrete implementations

  7. Property-Based Random Testing (12 tests)

  - Point-to-Point distance with 100 random test cases each
  - Point-to-LineString distance with random inputs
  - Point-to-Polygon distance with random inputs
  - Line-to-Line distance with random inputs
  - LineString-to-LineString distance with random inputs
  - Polygon-to-Polygon distance with random inputs
  - Line-to-Polygon distance with random inputs
  - LineString-to-Polygon distance with random inputs
  - Symmetry properties with random inputs
  - Edge cases and boundary conditions
  - Large coordinate values (1e6+ scale)
  - Small coordinate values (1e-6 scale)

  8. Geometric Edge Cases Tests (6 tests)

  - Collinear LineString geometries
  - Degenerate triangles (collinear points)
  - Self-intersecting polygons (bowtie/figure-8 shapes)
  - Nearly-touching geometries (1e-12 separation)
  - Very close but separate polygons (1e-14 gap)
  - Overlapping but non-intersecting LineStrings

  9. Numerical Precision Tests (3 tests)

  - Very close but non-zero distances (1e-15 to 1e-10 range)
  - Floating-point precision limits (near machine epsilon)
  - Large coordinate differences (mixed 1e-10 and 1e8 scales)

  10. Robustness Tests (6 tests)

  - NaN coordinate handling
  - Positive infinity coordinate handling
  - Negative infinity coordinate handling
  - Mixed special values (NaN + infinity combinations)
  - Subnormal (denormalized) number handling
  - Positive zero vs negative zero behavior

zhangfengcdt committed:
Fix the generic distance function and unit tests

zhangfengcdt committed:
add cross validation tests

zhangfengcdt committed:
add all tests cross validation

zhangfengcdt committed:
fix one failing test

zhangfengcdt committed:
Add cross validation to generic distance function and fix bugs

zhangfengcdt committed:
refacot to use the existing intersects function

zhangfengcdt committed:
refactor distance functions to utils.rs

zhangfengcdt committed:
revert distance.rs

zhangfengcdt committed:
Revert "revert distance.rs"

This reverts commit 12674145646cee321573ea843dc979766961cbe0.

zhangfengcdt committed:
fix clippy

zhangfengcdt committed:
clean up comments

zhangfengcdt committed:
refactor GenericDistanceTrait

zhangfengcdt committed:
more refactoring

zhangfengcdt committed:
cleanup unused codes

zhangfengcdt committed:
more cross-type distance

zhangfengcdt committed:
add impl_generic_distance_for_geometry_and_variant_new

zhangfengcdt committed:
macro - wip 1 - cord and point

zhangfengcdt committed:
wip - 2 point

zhangfengcdt committed:
wip 3 - line

zhangfengcdt committed:
wip 4 - LineString

zhangfengcdt committed:
add polygon

zhangfengcdt committed:
imple polygon

zhangfengcdt committed:
Rect and Triangle

zhangfengcdt committed:
wip

zhangfengcdt committed:
refactor GenericDistanceTrait

zhangfengcdt committed:
refactor tests to use the DistanceExt directly

zhangfengcdt committed:
fix multi-geometry types

zhangfengcdt committed:
refactor Implementation for Geometry (generic traits)

zhangfengcdt committed:
finish line

zhangfengcdt committed:
implement linestring

zhangfengcdt committed:
implement polygon

zhangfengcdt committed:
add triangle and rect

zhangfengcdt committed:
all types other than collection type

zhangfengcdt committed:
implement all types in macro

zhangfengcdt committed:
support same type GeometryCollection

zhangfengcdt committed:
rearrange

zhangfengcdt committed:
replace the dispatcher with macro

zhangfengcdt committed:
handle GeometryCollection

zhangfengcdt committed:
fix the ci tests

zhangfengcdt committed:
add generic trait for GeometryCollection

zhangfengcdt committed:
more edit

zhangfengcdt committed:
build successfully

zhangfengcdt committed:
add unit tests to utils.rs

zhangfengcdt committed:
add random tests

zhangfengcdt committed:
clean up comments

zhangfengcdt committed:
add more tests

zhangfengcdt committed:
fix fmt issue

zhangfengcdt committed:
fix clippy error

Kontinuation commented on geo-generic-alg/src/algorithm/line_measures/metric_spaces/euclidean/utils.rs:
```diff
@@ -0,0 +1,1774 @@
+                    return mindist;
+                }
+            }
+        }
```
I'm not sure whether this will have better performance than the original geo's implementation or not, as we are creating lots of temporary geo_types objects during the distance computation.

zhangfengcdt replied:
Really good observation! I added some targeted benchmarks to compare polygon-to-polygon performance between concrete and generic algorithms and yes it did show a little slower (2%-20%). Therefore I did some optimization to the generic polygon-to-polygon distance function, so that:

  - Skip object creation when no holes present
  - Only create LineStrings when actually needed
  - Check for interiors before any allocations
  - Reuse generic trait methods directly

And it looks much better now.

Before the fix:

```
  - Simple polygons: Generic ~20% slower (787ns vs 655ns)
  - Polygons with holes: Generic ~10% slower (1.08μs vs 988ns)
  - Complex polygons: Generic ~2% slower (8.18μs vs 7.99μs)
```

After the fix:

```
  - Simple polygons: 66% faster (788ns → 262ns)
  - Polygons with holes: 6% faster (1.08μs → 1.02μs)
  - Complex polygons: 41% faster (8.18μs → 4.80μs)
```

Run it using:
cargo bench -p geo-generic-alg --bench distance

and check cases after generic_vs_concrete_polygon_...

Kontinuation reviewed (approved):
I believe that this should be correct. However, I still have some concerns regarding the performance of this generic implementation especially for the polygon-polygon case.

zhangfengcdt committed:
optimize distance_polygon_to_polygon_generic function

zhangfengcdt committed:
Some more optimizations using early intersects check early stop

Kontinuation commented on ADDITIONAL_OPTIMIZATIONS.md:
```diff
@@ -0,0 +1,94 @@
+# Additional Performance Optimizations for utils.rs
```
I have approved it before, but I found that there's an unrelated file here, so I'll update the review state to "Request changes"

zhangfengcdt replied:
It is removed from the codebase now.

zhangfengcdt committed:
remove helper md file

Kontinuation commented on geo-generic-alg/src/algorithm/line_measures/metric_spaces/euclidean/distance.rs:
```diff
@@ -1239,47 +1632,105 @@ mod tests {
+            let generic_dist1 = p.distance_ext(&v1);
+            let generic_dist2 = p.distance_ext(&v2);
+            let generic_min_dist = generic_dist1.min(generic_dist2);
+            assert_relative_eq!(generic_min_dist, 63.25345840347388);
```
Why not directly invoke distance_ext to compute the distance between the multilienstring and point?

Kontinuation commented on geo-generic-alg/src/algorithm/line_measures/metric_spaces/euclidean/distance.rs:
```diff
@@ -1294,31 +1745,95 @@ mod tests {
+                .iter()
+                .map(|point| point_distance_generic(&p, point))
+                .fold(Float::max_value(), |acc: f64, dist| acc.min(dist));
+            assert_relative_eq!(generic_min_dist, 64.03124237432849);
```
Why not directly compute point to multipoint distance using distance_ext? We are basically duplicating the internal logic of multipoint-point distance computation.

Kontinuation reviewed (commented):
I have picked some of the implementations of distance_ext (for instance, line to polygon) and found that the generic implementation has fundamental algorithmic changes compared to the non-generic version. I hope that everything is implemented correctly.

I have also found some strange unit test code that replicates the logic of distance_ext function, would you like to curate the implementation and tests to make sure that they are all written appropriately like what a seasoned human engineer will do?

zhangfengcdt committed:
change some unit tests to directly test on distance_ext

zhangfengcdt committed:
fix distance_linestring_to_polygon_generic

zhangfengcdt committed:
fix distance_polygon_to_polygon_generic

zhangfengcdt committed:
fix distance_point_to_linestring_generic

zhangfengcdt committed:
make line_segment_distance_generic use the same mathematic equations

zhangfengcdt committed:
add comments to utils.rs

zhangfengcdt commented:
@Kontinuation 
I have checked and compared the generic vs concrete implementations one by one in utils.rs and they should all matched together with identical or equivalent algorithms now. I also adjusted the order of the functions to match the concrete distance codebase (in distance.rs) with extensive comments in each function. There are minor differences due to the optimizations I added in line_to_polygon and polygon_to_polygon, but I don't think they would cause different results - this is also verified in the targeted unit tests in utils.rs. I explain them in the comments as well.

I also checked the unit tests and I have fixed those not direct testing on distance_ext already. I believe they are all good now.

Kontinuation commented on geo-generic-alg/src/algorithm/line_measures/metric_spaces/euclidean/utils.rs:
```diff
@@ -208,22 +438,47 @@ where
+        let first_point_inside = if has_holes {
+            // Check if first point of LineString is inside the exterior ring
+            if let Some(first_coord) = linestring.coords().next() {
+                // Simple point-in-polygon test using ray casting
+                let point_x = first_coord.x();
+                let point_y = first_coord.y();
+                let mut inside = false;
+                let ring_coords: Vec<_> = exterior.coords().collect();
+                let n = ring_coords.len();
+
+                if n > 0 {
+                    let mut j = n - 1;
+                    for i in 0..n {
+                        let xi = ring_coords[i].x();
+                        let yi = ring_coords[i].y();
+                        let xj = ring_coords[j].x();
+                        let yj = ring_coords[j].y();
+
+                        if ((yi > point_y) != (yj > point_y))
+                            && (point_x < (xj - xi) * (point_y - yi) / (yj - yi) + xi)
+                        {
+                            inside = !inside;
+                        }
+                        j = i;
+                    }
                 }
```
We directly call `ring_contains_coord` in other places but implement a ray-casting algorithm here from scratch. Would it be better to implement a generic trait-based ring_contains_coord? The main dependency of this function, `coord_pos_relative_to_ring`, is already generic.

Kontinuation reviewed (approved):
LGTM. The minor concerns could be addressed in future PRs. We can merge it for now.
